src/meds_torch/models/components/transformer_decoder.py

```diff
@@ -1,15 +1,286 @@
+from collections.abc import Callable
+from functools import wraps
+
 import torch
+from loguru import logger
 from omegaconf import DictConfig
-from torch import nn
+from torch.functional import F
+from x_transformers import TransformerWrapper
+from x_transformers.autoregressive_wrapper import (
+    FILTER_LOGITS_FN,
+    align_right,
+    exists,
+    identity,
+    join,
+    pack,
+    unpack,
+)
 
 from meds_torch.input_encoder import INPUT_ENCODER_MASK_KEY, INPUT_ENCODER_TOKENS_KEY
 from meds_torch.models import BACKBONE_EMBEDDINGS_KEY, BACKBONE_TOKENS_KEY
 from meds_torch.models.components.utils import get_last_token
 from meds_torch.utils.module_class import Module
 
 
+def eval_decorator(fn):
+    @wraps(fn)
+    def inner(model, *args, **kwargs):
+        was_training = model.training
+        model.eval()
+        out = fn(model, *args, **kwargs)
+        model.train(was_training)
+        return out
+
+    return inner
+
+
+class GenerationBudget:
+    """A class to handle generation budgets with mutually exclusive sequence length or time length.
+
+    This class ensures that only one of max_seq_len or min_time_len can be set.
+
+    Examples:
+        >>> # Create from sequence length
+        >>> budget_seq = GenerationBudget.from_seq_len(100)
+        >>> budget_seq.max_seq_len
+        100
+        >>> budget_seq.min_time_len is None
+        True
+
+        >>> # Create from time length
+        >>> budget_time = GenerationBudget.from_time_len(60)
+        >>> budget_time.min_time_len
+        60
+        >>> budget_time.max_seq_len is None
+        True
+
+        >>> # Values remain consistent
+        >>> budget_seq = GenerationBudget.from_seq_len(500)
+        >>> (budget_seq.max_seq_len, budget_seq.min_time_len)
+        (500, None)
+
+        >>> # Direct initialization isn't intended for use
+        >>> budget = GenerationBudget(100)
+    """
+
+    def __init__(self, value: int):
+        self._value = value
+
+    @classmethod
+    def from_seq_len(cls, value: int) -> "GenerationBudget":
+        """Create a GenerationBudget from a maximum sequence length.
+
+        Args:
+            value: The maximum sequence length
+
+        Returns:
+            A GenerationBudget instance with max_seq_len set
+        """
+        instance = cls(value)
+        instance._is_seq_len = True
+        return instance
+
+    @classmethod
+    def from_time_len(cls, value: int) -> "GenerationBudget":
+        """Create a GenerationBudget from a minimum time length.
```

Comment:
For functions this simple a one-line docstring is acceptable. https://google.github.io/styleguide/pyguide.html#383-functions-and-methods